nixos/modules/services/networking/tuntox.nix

```diff
@@ -0,0 +1,108 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.tuntox.server;
+  hostsWhitelistFile = pkgs.writeText "tuntox-hosts-whitelist" ''
+    ${toString cfg.hostsWhitelist}
+  '';
+  persistentIdDirectory = "/var/lib/tuntox";
+in
+{
+  options = {
+    services.tuntox.server = {
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = "If enabled, start the Tuntox Service.";
+      };
+
+      relayPort = mkOption {
+        type = types.nullOr types.string;
+        default = null;
+        description = "Set TCP relay port (0 disables TCP relaying)";
+      };
+
+      portRange = mkOption {
+        type = types.nullOr types.string;
+        default = null;
+        description = "Set Tox UDP port range (<literal>port:port</literal>)";
+      };
+
+      persistentId = mkOption {
+        type = types.bool;
+        default = false;
+        description = "If enabled, persist the tox id between runs";
+      };
+
+      hostsWhitelist = mkOption {
+        type = types.listOf (types.separatedString "\n");
+        default = [];
+        description = "Only allow connections to the specified <literal>hostname:port</literal> pairs";
+      };
+
+      idWhitelist = mkOption {
+        type = types.listOf (types.separatedString " ");
+        default = [];
+        description = "Whitelisted Tox IDs. If empty, id whitelisting will be disabled";
+      };
+
+      secret = mkOption {
+        type = types.nullOr types.str;
+        default = null;
+        description = "Shared secret used for connection authentication";
```

Comment:
It might be good to point out that this secret will be world readable when inserted in the Nix store. Can this optionally be done with a secretFile that is stored outside the nix store?